### DISCUSSION TITLE
Issue with running a merge involving a collection class

### DISCUSSION THREAD

**Discussion by lukeschissler** [sqlalchemy/sqlalchemy#11842] [Usage Questions]
Seeing this error when attempting a merge on a an ORM object which contains a collection class - not seeing a lot of info for this error online so asking here.

![Screenshot 2024-09-05 at 10 36 23 AM](https://github.com/user-attachments/assets/c876a4ac-9d5a-42e3-ab37-ece0ee594177)

The ORM schema in question

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String, nullable=False)
    signal_map_id: Mapped[int] = mapped_column(
        ForeignKey("signal_maps.id"), nullable=False
    )
    created_by: Mapped[str] = mapped_column(String(40), nullable=True)
    device_wrapper_id: Mapped[int] = mapped_column(
        ForeignKey("device_wrappers.id", name="signal_device_wrapper_id_fk"),
        nullable=False,
    )
    device_id: Mapped[int] = mapped_column(
        ForeignKey("devices.id", name="signal_device_uuid_fk"), nullable=True
    )
    device: Mapped[Device] = relationship()
    device_wrapper: Mapped[DeviceWrapper] = relationship()

    async def get_device(self) -> Device:
        if self.device_id:
            return cast(Device, await self.awaitable_attrs.device)
        return cast(
            Device,
            await cast(
                DeviceWrapper, await self.awaitable_attrs.device_wrapper
            ).awaitable_attrs.current_version,
        )

    measurement_type_id: Mapped[int] = mapped_column(
        ForeignKey("measurement_types.id", name="signal_measurement_type_fk"),
        nullable=False,
    )
    measurement_type: Mapped["MeasurementType"] = relationship(
        foreign_keys=[measurement_type_id]
    )
    signal_metadata: Mapped[dict[str, "SignalMetadata"]] = relationship(
        back_populates="signal",
        collection_class=attribute_keyed_dict("field_name"),
        cascade="all, delete-orphan",
    )
    signal_groups: Mapped[List["SignalGroup"]] = relationship(
        back_populates="signal", cascade="all, delete-orphan"
    )
    signal_map: Mapped[SignalMap] = relationship(back_populates="signals")`

Any help would be appreciated, thanks!

**Comment by zzzeek** [comment]
it's raising on this:

```
if self[key] != value:
```

Which looks a little bit like not how we usually do this on our part, we aren't supposed to call `__eq__()` on user-mapped objects.  But what it's trying to do is determine that the object that was asked to be removed from the dictionary is the same object that we actually found in the dictionary.   so the merge call here is asking an object to be removed that matches the same "key" but is not the same object.

I dont really know offhand how to get merge to do that.    here's a demonstration.  can you alter this demo to resemble your own code enough that it shows the same error?  the async stuff should not be needed, thanks

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy.orm import attribute_keyed_dict
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    bs: Mapped[list[B]] = relationship(
        "B",
        collection_class=attribute_keyed_dict("key"),
        cascade="all, delete-orphan",
    )


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))
    key: Mapped[str]


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(id=1, bs={"b1": B(id=1, key="b1"), "b2": B(id=2, key="b2")}))
s.commit()


s = Session(e)

a1 = A(id=1, bs={"b1": B(id=1, key="b1")})
a2 = s.merge(a1)
s.flush()

```
